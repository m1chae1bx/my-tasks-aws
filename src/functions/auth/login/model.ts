import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface LoginRequest {
  body: {
    id: string;
    password: string;
  };
}

export const loginRequestSchema: JSONSchemaType<LoginRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    body: {
      type: "object",
      properties: {
        id: {
          type: "string",
          maxLength: 100,
        },
        password: {
          type: "string",
        },
      },
      required: ["id", "password"],
    },
  },
  required: ["body"],
};

export const validateRequest = ajv.compile(loginRequestSchema);
