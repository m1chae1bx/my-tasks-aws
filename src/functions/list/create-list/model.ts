import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface CreateListRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    name: string;
    isDefault?: boolean;
  };
}

export const createListRequestSchema: JSONSchemaType<CreateListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 100,
        },
        isDefault: {
          type: "boolean",
          nullable: true,
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createListRequestSchema);
