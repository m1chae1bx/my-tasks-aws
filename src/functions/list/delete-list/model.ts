import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteListRequest {
  pathParameters: {
    userId: string;
    listId: string;
  };
}

export const deleteListRequestSchema: JSONSchemaType<DeleteListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
        listId: {
          type: "string",
        },
      },
      required: ["userId", "listId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteListRequestSchema);
