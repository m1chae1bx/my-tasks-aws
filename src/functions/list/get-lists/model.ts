import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface GetListsRequest {
  pathParameters: {
    userId: string;
  };
}

export const getListsRequestSchema: JSONSchemaType<GetListsRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getListsRequestSchema);
