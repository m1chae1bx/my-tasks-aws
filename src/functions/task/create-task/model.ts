import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface CreateTaskRequest {
  pathParameters: {
    listId: string;
  };
  body: {
    name: string;
    desc?: string;
    dueDate?: string;
  };
}

export const createTaskRequestSchema: JSONSchemaType<CreateTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
      },
      required: ["listId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 200,
        },
        desc: {
          type: "string",
          nullable: true,
          maxLength: 400,
        },
        dueDate: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createTaskRequestSchema);
