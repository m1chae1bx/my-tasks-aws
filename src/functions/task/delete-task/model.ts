import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteTaskRequest {
  pathParameters: {
    listId: string;
    taskId: string;
  };
}

export const deleteTaskRequestSchema: JSONSchemaType<DeleteTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
        taskId: {
          type: "string",
        },
      },
      required: ["listId", "taskId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteTaskRequestSchema);
