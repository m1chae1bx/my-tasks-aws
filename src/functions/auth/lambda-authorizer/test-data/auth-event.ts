import { baseTokenAuthorizerEvent } from "@test-utils/base-objects";

export const authEvent = baseTokenAuthorizerEvent;
