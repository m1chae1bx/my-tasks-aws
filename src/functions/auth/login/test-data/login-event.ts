import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const loginEvent = {
  ...baseAPIGatewayProxyEvent,
  body: JSON.stringify({
    id: "testUserId",
    password: "secret-password",
  }),
};
