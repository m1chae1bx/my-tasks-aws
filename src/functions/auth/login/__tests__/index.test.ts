import { handler } from "../index";
import { loginEvent } from "../test-data/login-event";
import { validateRequest } from "../model";
import { User } from "@libs/user";
import { APIGatewayProxyResult } from "aws-lambda";
import { ajvError } from "@test-utils/base-objects";
import { genericErrorHandler } from "@libs/generic/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("@libs/user", () => ({
  User: { get: jest.fn() },
}));

jest.mock("@libs/generic/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("login", () => {
  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce({
        validatePassword: jest.fn().mockReturnValueOnce(true),
        generateJwt: jest.fn().mockReturnValueOnce("test-token"),
      });
      result = await handler(loginEvent);
    });

    it("should return 200 status code", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return token", async () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain("test-token");
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler(loginEvent);
    });

    it("should return 400 status code", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return errors", () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain(ajvError[0].message);
    });
  });

  describe("sad path - invalid request (no body)", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler({ ...loginEvent, body: null });
    });

    it("should return 400 status code", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return errors", () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain(ajvError[0].message);
    });
  });

  describe("sad path - incorrect password", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce({
        validatePassword: jest.fn().mockReturnValueOnce(false),
      });
      result = await handler(loginEvent);
    });

    it("should return 401 status code", () => {
      expect(result.statusCode).toBe(401);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - user not found", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce(null);
      result = await handler(loginEvent);
    });

    it("should return 401 status code", () => {
      expect(result.statusCode).toBe(401);
    });
  });

  describe("sad path - generic error", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockRejectedValueOnce(new Error("test error"));
      (genericErrorHandler as jest.Mock).mockReturnValueOnce({
        statusCode: 500,
        body: "test error",
      });
      result = await handler(loginEvent);
    });

    it("should return 500 status code", () => {
      expect(result.statusCode).toBe(500);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });
});
