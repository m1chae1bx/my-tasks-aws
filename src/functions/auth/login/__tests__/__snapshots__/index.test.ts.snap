// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login happy path should return token 1`] = `"{\\"token\\":\\"test-token\\"}"`;

exports[`login sad path - generic error should return an error message 1`] = `"test error"`;

exports[`login sad path - incorrect password should return an error message 1`] = `"{\\"message\\":\\"Password is incorrect\\"}"`;

exports[`login sad path - invalid request should return errors 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;
